### PR TITLE
complete task queue traits

### PR DESCRIPTION
**Pull request by sticnarf**
* Remove the trait function creating the queue. Instead, use free functions to create the pair of `TaskInjector` and `LocalQueue`s.
* Add `LocalQueue` trait.
* `LocalQueue::pop` returns the instant when the task was pushed to the queue and whether the task is from the current local queue.

### CODE REVIEW

File: src/queue/mod.rs [hunk 1]
```diff
@@ -16,18 +20,25 @@ pub trait TaskCell {
...
-    /// If the queue is closed, the method should behave like no-op.
     fn push(&self, task: Self::TaskCell);
+}
+
+/// The consumer of a task queue.
+pub trait Consumer {
+    /// The task cell in the queue.
+    type TaskCell: TaskCell;
 
-    /// Closes the queue so that no more tasks can be scheduled.
```
Comment (BusyJay) [inline src/queue/mod.rs]: Why remove it?

Comment (sticnarf) [reply src/queue/mod.rs]: What's the purpose of it?

Comment (BusyJay) [reply src/queue/mod.rs]: As the comment says, close the queue so that no more tasks can be scheduled. See also crossbeam-rs/crossbeam#236.

Comment (sticnarf) [reply src/queue/mod.rs]: Should it be the thread pool who prevents the tasks from being schduled?

Comment (BusyJay) [reply src/queue/mod.rs]: How can it prevent scheduling by `Remote`?

Comment (BusyJay) [reply src/queue/mod.rs]: What I thought is to reduce an atomic load by handling it inside the queue. But if level queue can't benefit from it, I also feel OK to do it in the pool part.

Comment (sticnarf) [reply src/queue/mod.rs]: I'm curious how it reduces an atomic load? The closed flag  should be shared among all senders. Change the inner code of `crossbeam_deque::Injector::push`?

Comment (BusyJay) [reply src/queue/mod.rs]: It depends on the implement details. For the original queue implement, the desired way is to change the code of `Injector`.

File: src/queue/mod.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+/// The local queue of a task queue.
+pub trait LocalQueue {
+    /// The task cell in the queue.
+    type TaskCell: TaskCell;
+
+    /// Pushes a task to the local queue.
+    fn push(&mut self, task_cell: Self::TaskCell);
+
+    /// Gets a task cell from the queue. Rturns `None` if there is no task cell
```
Comment (BusyJay) [inline src/queue/mod.rs]: s/Rturns/Returns/

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: src/queue/mod.rs (hunk 2 above)
Comment (sticnarf) [reply src/queue/mod.rs]: Fixed.

**Review by BusyJay — APPROVED** [review]

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**